include/nuttx/tee.h

```diff
@@ -0,0 +1,479 @@
+
+#define TEE_IOCTL_LOGIN_REE_KERNEL              0x80000000
+
+/**
+ * struct tee_ioctl_param - parameter
+ * @attr: attributes
+ * @a: if a memref, offset into the shared memory object, else a value
+ * parameter
+ * @b: if a memref, size of the buffer, else a value parameter
+ * @c: if a memref, shared memory identifier, else a value parameter
+ *
+ * @attr & TEE_PARAM_ATTR_TYPE_MASK indicates if memref or value is used in
+ * the union. TEE_PARAM_ATTR_TYPE_VALUE_* indicates value and
+ * TEE_PARAM_ATTR_TYPE_MEMREF_* indicates memref. TEE_PARAM_ATTR_TYPE_NONE
+ * indicates that none of the members are used.
+ *
+ * Shared memory is allocated with TEE_IOC_SHM_ALLOC which returns an
+ * identifier representing the shared memory object. A memref can reference
+ * a part of a shared memory by specifying an offset (@a) and size (@b) of
+ * the object. To supply the entire shared memory object set the offset
+ * (@a) to 0 and size (@b) to the previously returned size of the object.
+ */
+
+struct tee_ioctl_param
+{
+  uint64_t attr;
+  uint64_t a;
+  uint64_t b;
+  uint64_t c;
+};
+
+#define TEE_IOCTL_UUID_LEN  16
+
+/**
+ * struct tee_ioctl_open_session_arg - Open session argument
+ * @uuid:       [in] UUID of the Trusted Application
+ * @clnt_uuid:  [in] UUID of client
+ * @clnt_login: [in] Login class of client, TEE_IOCTL_LOGIN_* above
+ * @cancel_id:  [in] Cancellation id, a unique value to identify this request
+ * @session:    [out] Session id
+ * @ret:        [out] return value
+ * @ret_origin  [out] origin of the return value
+ * @num_params  [in] number of parameters following this struct
+ */
+
+struct tee_ioctl_open_session_arg
+{
+  uint8_t uuid[TEE_IOCTL_UUID_LEN];
+  uint8_t clnt_uuid[TEE_IOCTL_UUID_LEN];
+  uint32_t clnt_login;
+  uint32_t cancel_id;
+  uint32_t session;
+  uint32_t ret;
+  uint32_t ret_origin;
+  uint32_t num_params;
+
+  /* num_params tells the actual number of element in params */
+
+  struct tee_ioctl_param params[];
+};
+
+/**
+ * TEE_IOC_OPEN_SESSION - opens a session to a Trusted Application
+ *
+ * Takes a struct tee_ioctl_buf_data which contains a struct
+ * tee_ioctl_open_session_arg followed by any array of struct
+ * tee_ioctl_param
+ */
+
+#define TEE_IOC_OPEN_SESSION _IOC(TEE_IOC_MAGIC << 8, TEE_IOC_BASE + 2)
+
+/**
+ * struct tee_ioctl_invoke_func_arg - Invokes a function in a Trusted
+ * Application
+ * @func:       [in] Trusted Application function, specific to the TA
+ * @session:    [in] Session id
+ * @cancel_id:  [in] Cancellation id, a unique value to identify this request
+ * @ret:        [out] return value
+ * @ret_origin  [out] origin of the return value
+ * @num_params  [in] number of parameters following this struct
+ */
+
+struct tee_ioctl_invoke_arg
+{
+  uint32_t func;
+  uint32_t session;
+  uint32_t cancel_id;
+  uint32_t ret;
+  uint32_t ret_origin;
+  uint32_t num_params;
+
+  /* num_params tells the actual number of element in params */
+
+  struct tee_ioctl_param params[];
```

Comment:
C89 incompatible

Reply:
This is a standard interface that cannot be changed and needs to be compatible with the optee client data structure in user space.

Reply:
> This is a standard interface that cannot be changed and needs to be compatible with the optee client data structure in user space.

Is there a workaround to make it C89-compatible? e.g., array of 1, array of 2, or array of fixed size?

Reply:
> This interface is defined by open optee client, and the whole software stack doesn't support C89. It doesn't make sense to require driver support C89, but the userspace library not.

That is annoying, but I guess nothing we can do about it now.